### ISSUE TITLE
Python Detection inside VirtualEnv Picks Wrong Library (Trac #2367)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2359]
When calling cmake inside a Python virtualenv, a wrong python library might be found.

On MacOS 10.15 I run:

```text
brew install python@2
pip2 install --upgrade pip setuptools
pip2 install virtualenv
python2 -m virtualenv EnvPy2
```

Then I install numpy, scipy, tables, etc. inside that virtual environment.
Now, on combo/trunk, cmake shows:

```text
-- python
-- +  version: 2.7.16
-- + base dir: /Users/thomas/IceCube/EnvPy2/bin/..
-- +   binary: /Users/thomas/IceCube/EnvPy2/bin/python
-- + includes: /Users/thomas/IceCube/EnvPy2/include/python2.7
-- +     libs: /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/lib/libpython2.7.tbd
-- +    numpy: /Users/thomas/IceCube/EnvPy2/lib/python2.7/site-packages/numpy/core/include
```
The issue: It picks the correct python, that is linked in the virtualenv, but then takes the libpython from the system python.

In `cmake/tools/python.cmake` we have:

```text
#
# Get the root dir of the python install
#
execute_process(COMMAND ${PYTHON_EXECUTABLE} -c "import sys; sys.stdout.write(sys.prefix)"
                OUTPUT_VARIABLE PYTHON_ROOT)
message(STATUS "+ base dir: ${PYTHON_ROOT}")

#
# Find the library and header file manually,
# using the python root that we found above.
#
FIND_LIBRARY(PYTHON_LIBRARY
  NAMES python${PYTHON_VERSION_NO_DOTS} python${PYTHON_STRIPPED_MAJOR_MINOR_VERSION}
  PATHS ${PYTHON_ROOT} ${PYTHON_ROOT}/lib
  PATH_SUFFIXES
    python${PYTHON_STRIPPED_MAJOR_MINOR_VERSION}
  NO_SYSTEM_ENVIRONMENT_PATH
  NO_DEFAULT_PATH
)
```

When running inside a virtualenv, `sys.prefix` points to the venv and `FIND_LIBRARY` cannot find the library in there. We should use `sys.real_prefix`, which points to the base directory:

```text
>>> sys.prefix
'/Users/thomas/IceCube/EnvPy2/bin/..'
>>> sys.real_prefix
'/usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7'
```

That fixes the issue for me:

```text
-- python
-- +  version: 2.7.16
-- + base dir: /usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7
-- +   binary: /Users/thomas/IceCube/EnvPy2/bin/python
-- + includes: /usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7/include/python2.7
-- +     libs: /usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7/lib/libpython2.7.dylib
```

Migrated from https://code.icecube.wisc.edu/ticket/2367
```json
{
    "status": "closed", 
    "changetime": "2019-11-05T21:14:10", 
    "description": "When calling cmake inside a Python virtualenv, a wrong python library might be found.\n\nOn MacOS 10.15 I run:\n{{{\nbrew install python@2\npip2 install --upgrade pip setuptools\npip2 install virtualenv\npython2 -m virtualenv EnvPy2\n}}}\n\nThen I install numpy, scipy, tables, etc. inside that virtual environment.\nNow, on combo/trunk, cmake shows:\n{{{\n-- python\n-- +  version: 2.7.16\n-- + base dir: /Users/thomas/IceCube/EnvPy2/bin/..\n-- +   binary: /Users/thomas/IceCube/EnvPy2/bin/python\n-- + includes: /Users/thomas/IceCube/EnvPy2/include/python2.7\n-- +     libs: /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/lib/libpython2.7.tbd\n-- +    numpy: /Users/thomas/IceCube/EnvPy2/lib/python2.7/site-packages/numpy/core/include\n}}}\nThe issue: It picks the correct python, that is linked in the virtualenv, but then takes the libpython from the system python.\n\nIn `cmake/tools/python.cmake` we have:\n{{{\n#\n# Get the root dir of the python install\n#\nexecute_process(COMMAND ${PYTHON_EXECUTABLE} -c \"import sys; sys.stdout.write(sys.prefix)\"\n                OUTPUT_VARIABLE PYTHON_ROOT)\nmessage(STATUS \"+ base dir: ${PYTHON_ROOT}\")\n\n#\n# Find the library and header file manually,\n# using the python root that we found above.\n#\nFIND_LIBRARY(PYTHON_LIBRARY\n  NAMES python${PYTHON_VERSION_NO_DOTS} python${PYTHON_STRIPPED_MAJOR_MINOR_VERSION}\n  PATHS ${PYTHON_ROOT} ${PYTHON_ROOT}/lib\n  PATH_SUFFIXES\n    python${PYTHON_STRIPPED_MAJOR_MINOR_VERSION}\n  NO_SYSTEM_ENVIRONMENT_PATH\n  NO_DEFAULT_PATH\n)\n}}}\n\nWhen running inside a virtualenv, `sys.prefix` points to the venv and `FIND_LIBRARY` cannot find the library in there. We should use `sys.real_prefix`, which points to the base directory:\n{{{\n>>> sys.prefix\n'/Users/thomas/IceCube/EnvPy2/bin/..'\n>>> sys.real_prefix\n'/usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7'\n}}}\n\nThat fixes the issue for me:\n{{{\n-- python\n-- +  version: 2.7.16\n-- + base dir: /usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7\n-- +   binary: /Users/thomas/IceCube/EnvPy2/bin/python\n-- + includes: /usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7/include/python2.7\n-- +     libs: /usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7/lib/libpython2.7.dylib\n}}}", 
    "reporter": "thomas.kintscher", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1572988450788714", 
    "component": "cmake", 
    "summary": "Python Detection inside VirtualEnv Picks Wrong Library", 
    "priority": "normal", 
    "keywords": "", 
    "time": "2019-10-21T13:04:52", 
    "milestone": "", 
    "owner": "thomas.kintscher", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20191021T13:05:32: 
* thomas.kintscher changed status from "new" to "closed"
* thomas.kintscher commented:

> In changeset 176286:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="176286"
> Fix python detection in virtualenv. Fixes #2359.
> ```


* thomas.kintscher changed owner from "" to "thomas.kintscher"
* thomas.kintscher changed resolution from "" to "fixed"

**Comment by icecube-trac** [comment]
Trac update at 20191021T13:40:39: thomas.kintscher commented:

> In changeset 176288:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="176288"
> Fix Python detection inside venv also for Python3. Fixes #2359 also for Python3.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20191105T21:10:48: 
* thomas.kintscher commented:

> The proposed change breaks non-virtualenv Python2.
> In Python2 inside the virtualenv, it is `sys.real_prefix`.
> Outside the virtualenv, it is `sys.prefix` and `sys.real_prefix` does not exist. m(


* thomas.kintscher changed status from "closed" to "reopened"
* thomas.kintscher changed resolution from "fixed" to ""

**Comment by icecube-trac** [comment]
Trac update at 20191105T21:14:10: 
* thomas.kintscher commented:

> In changeset 176786:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="176786"
> Fix Python2 detection outside of virtualenvs. Fixes #2359 for real, I hope.
> ```


* thomas.kintscher changed resolution from "" to "fixed"
* thomas.kintscher changed status from "reopened" to "closed"
